SelectionDAG/expandFMINIMUMNUM_FMAXIMUMNUM: FCANONICALIZE is needed only for sNaN

wzssyqa opened the pull request:
If we are sure that it is not sNaN, even it may be qNaN, we can use it directly.

wzssyqa committed:
SelectionDAG/expandFMINIMUMNUM_FMAXIMUMNUM: FCANONICALIZE is needed only for sNaN

If we are sure that it is not sNaN, even it may be qNaN, we can use
it directly.

arsenm reviewed (commented):
Needs test changes

wzssyqa commented:
> Needs test changes

I only see that we have `nnan` option while there is no similiar one for sNaN.
https://llvm.org/docs/LangRef.html

How to mark a variable to never SNaN while may qNaN?

wzssyqa commented:
I guess t hat I am confused by `nofpclass` in https://llvm.org/docs/LangRef.html#id1932.
Can it used to mark an argument that it will never be a type?

I have a try, while it won't as `DAG.isKnownNeverNaN` won't return true in this case.
Maybe I misunderstand its meaning.

wzssyqa commented:
> Needs test changes
Once https://github.com/llvm/llvm-project/pull/108350 is merged, I will fix SelectionDAG::isKnownNeverNaN with sNaN, and add test case for it.

arsenm commented:
> I guess t hat I am confused by `nofpclass` in https://llvm.org/docs/LangRef.html#id1932. Can it used to mark an argument that it will never be a type?
> 
> I have a try, while it won't as `DAG.isKnownNeverNaN` won't return true in this case. Maybe I misunderstand its meaning.

isKnownNeverNaN is quite underdeveloped. You can most easily test the no-nan operand case by just having a filler fadd nnan instruction. It would be simpler to do that than pulling in a dependency on some more general nofpclass handling

wzssyqa committed:
SelectionDAG: Support nofpclass(nan/qnan/snan)

SelectionDAGISel::LowerArguments: Pass NoNaN Flags to InVals.

`nofpclass` support values nan, snan, qnan, where nan=snan|qnan.
So let's use NoSNaNs and NoQNaNs in SDNodeFlags.

Thus, we can use it in isKnownNeverNaN.
